VALIDATOR_GUIDE.md

```diff
@@ -0,0 +1,158 @@
+    v, r, s = utils.ecdsa_raw_sign(sighash, key)
+    sig = utils.encode_int32(v) + utils.encode_int32(r) + utils.encode_int32(s)
+    return rlp.encode([validator_index, target_hash, target_epoch, source_epoch, sig])
+```
+
+## Logout Message Generation
+Like the Casper vote messages, a logout message is an RLP encoded list where the last element is the validator's signature. The elements of the unsigned list are the `validator_index` and `epoch` where epoch is the current epoch. A signature is generated in the same way it is done with votes above.
+
+This is [implemented in Pyethereum](https://github.com/karlfloersch/pyethereum/blob/develop/ethereum/hybrid_casper/casper_utils.py#L79-L83) as follows:
+```
+def mk_logout(validator_index, epoch, key):
+    sighash = utils.sha3(rlp.encode([validator_index, epoch]))
+    v, r, s = utils.ecdsa_raw_sign(sighash, key)
+    sig = utils.encode_int32(v) + utils.encode_int32(r) + utils.encode_int32(s)
+    return rlp.encode([validator_index, epoch, sig])
+```
+
+## Simple Casper Contract High Level Overview
+The Casper smart contract contains Casper's core logic. It is written in Viper &amp; can be deployed to the blockchain like any other contract to `CASPER_ADDR`. Casper messages are then sent to the contract by calling `vote(vote_msg)` where `vote_msg` is a Casper vote messaged as outlined above.
+
+### [[Contract Source]](https://github.com/ethereum/casper/blob/master/casper/contracts/simple_casper.v.py)
+
+#### `def __init__(epoch_length, withdrawal_delay, ...)`
+The Casper contract constructor takes in key settings. Most notably: 
+- `epoch_length`
+- `withdrawal_delay`
+- `dynasty_logout_delay`
+- `min_deposit_size`
+- `base_interest_factor`
+- `base_penalty_factor`
+
+These settings cannot be changed after deployment.
+
+#### `def initialize_epoch(epoch)`
+Calculates the interest rate &amp; penalty factor for this epoch based on the time since finality.
+
+Once a new epoch begins, this function is immediately called as the first transaction applied to the state. See [Epoch initialization](http://notes.eth.sg/GwTgLAZghgRgHGAtABmARgKaLFNzEIwjZgDsAxsBAMzIwbJpA===?both#epoch-initialization) for more details.
+
+#### `def deposit(validation_addr, withdrawal_addr)`
+Accepts deposits from prospective validators &amp; adds them to the next validator set.
+
+#### `def logout(logout_msg)`
+Initiates validator logout. The validator must continue to validate for `dynasty_logout_delay` dynasties before entering the `withdrawal_delay` waiting period.
+
+#### `def withdraw(validator_index)`
+If the validator has waited for a period greater than `withdrawal_delay` epochs past their `end_dynasty`, then send them ETH equivalent to their deposit.
+
+#### `def vote(vote_msg)`
+Called once by each validator each epoch. The vote message contains the fields presented in [Casper Vote Format](http://notes.eth.sg/GwTgLAZghgRgHGAtABmARgKaLFNzEIwjZgDsAxsBAMzIwbJpA===?both#casper-vote-format).
```

Comment:
some links are still referencing notes.eth.sg. Those are originally referencing to headers in the doc itself.

Reply:
thanks!